backend/editor/OT/testing_framework.go

```diff
@@ -0,0 +1,87 @@
+package editor
+
+import (
+	"cms.csesoc.unsw.edu.au/editor/OT/datamodel"
+	"cms.csesoc.unsw.edu.au/editor/OT/operations"
+	"cms.csesoc.unsw.edu.au/environment"
+	"cms.csesoc.unsw.edu.au/pkg/cmsjson"
+	"github.com/google/uuid"
+	"github.com/gorilla/websocket"
+)
+
+// testing_framework is a simple framework that can be used for performing integration tests on the concurrent editor
+// it allows for the observation of client behavior and tracking document server behavior
+type TestingClient struct {
+	underlyingClient *clientView
+	operationPipe    pipe
+	terminationPipe  func()
+}
+
+func (tC TestingClient) HasTerminated() bool   { return len(tC.underlyingClient.sendTerminateSignal) > 0 }
+func (tC TestingClient) WasAcknowledged() bool { return len(tC.underlyingClient.sendOp) > 0 }
```

Comment:
why is it `sendOp` and not `sendAcknowledgement` here?
also, would it always be false since [the condition in select-case](https://github.com/csesoc/website/blob/36cfebae51a6b16888af5bc152b5eb4e03b93e0b/backend/editor/OT/client_view.go#L39) i.e. `<-c.sendOp` takes the message out of the channel?

Reply:
Oooh yup thanks for the catch, yeah it should be `sendAcknowledgement` also `WasAcknowledged` will always be false until we get an acknowledgement and from then on-wards it will be true, the point of this framework is to sorta "emulate" that run function without actually running it

although we do probably need a way for the framework to acknowledge the ack so ill add that in 😎